pkg/kubelet/pod/mirror/manager.go

```diff
@@ -0,0 +1,372 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mirror
+
+import (
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	kubepod "k8s.io/kubernetes/pkg/kubelet/pod"
+	kubetypes "k8s.io/kubernetes/pkg/kubelet/types"
+	"k8s.io/kubernetes/pkg/kubelet/util/format"
+	"k8s.io/kubernetes/pkg/util"
+)
+
+// Mirror pod manager stores and manages all the mirror pods.
+//
+// Kubelet discovers pod updates from 3 sources: file, http, and apiserver.
+// Pods from non-apiserver sources are called static pods, and API server is
+// not aware of the existence of static pods. In order to monitor the status of
+// such pods, kubelet creates a mirror pod for each static pod via the API
+// server.
+//
+// A mirror pod has the same pod full name (name and namespace) as its static
+// counterpart (albeit different metadata such as UID, etc). By leveraging the
+// fact that kubelet reports the pod status using the pod full name, the status
+// of the mirror pod always reflects the actual status of the static pod.
+// When a static pod gets deleted, the associated orphaned mirror pod will
+// also be removed.
+
+type Manager interface {
+	// Add and Delete a static pod, and the corresponding mirror pod will be created, recreated
+	// and deleted.
+	AddStaticPod(staticPod *api.Pod)
+	DeleteStaticPod(staticPod *api.Pod)
+
+	// Check a mirror pod, recreate mirror pod when needed
+	CheckMirrorPod(mirrorPod *api.Pod)
+
+	// Get current mirror pod by pod full name, if there is no mirror pod now (it hasn't been
+	// created or recreated), return nil.
+	GetMirrorPod(podFullName string) *api.Pod
+
+	// Get all mirror pods.
+	// TODO: (random-liu) Add this to kubelet.GetPods(), because mesos needs this.
+	GetMirrorPods() []*api.Pod
+
+	// Check whether a mirror pod is valid
+	IsValidMirrorPod(mirrorPod *api.Pod) bool
+}
+
+type basicManager struct {
+	// Protects staticPods and invalidMirrorPods.
+	lock sync.RWMutex
+	// Static pods indexed by full name
+	// TODO: (random-liu) Remove staticPods, and rely on pod manager is also feasible.
+	staticPods map[string]*api.Pod
+	// Buffer of invalid mirror pods
+	// TODO: (random-liu) Invalid mirror pods can also be passed to manager goroutine by channel,
+	// however that may be a little more complicated, because the channel may be full. For now
+	// just deal with invalid mirror pods in periodically cleanup. May use channel in the future.
+	invalidMirrorPods map[string]*api.Pod
+
+	// Protects mirrorPods.
+	// mirrorPods will ONLY BE MODIFIED IN THE MANAGER GOROUTINE, and read from other goroutines.
+	// mirrorPods can be READ WITHOUT LOCK in the manager goroutine, because no other goroutines
+	// will modify it.
+	mirrorPodsLock sync.RWMutex
```

Comment:
The two different locks worry me a bit since one method may hold both locks at the same time. I checked your implementation and even though it's not possible to have deadlocks right now, it's hard to say in the future. How bad is it to use a single lock?